File: libres/lib/enkf/local_config.cpp
/*
   Copyright (C) 2011  Equinor ASA, Norway.

   The file 'local_config.c' is part of ERT - Ensemble based Reservoir Tool.

   ERT is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   ERT is distributed in the hope that it will be useful, but WITHOUT ANY
   WARRANTY; without even the implied warranty of MERCHANTABILITY or
   FITNESS FOR A PARTICULAR PURPOSE.

   See the GNU General Public License at <http://www.gnu.org/licenses/gpl.html>
   for more details.
*/

#include <stdlib.h>
#include <stdio.h>

#include <ert/util/vector.h>

#include <ert/enkf/local_ministep.hpp>
#include <ert/enkf/local_updatestep.hpp>
#include <ert/enkf/local_config.hpp>

#include "ert/python.hpp"
static auto logger = ert::get_logger("local_config");

/*

  +-------------------------- LocalUpdateStep ---------------------------------------+
  |                                                                                        |
  |                                                                                        |
  |    +----------------- local_ministep_type --------------------------------------+      |
  |    |                                                                            |      |
  |    |                                       /    +--- local_dataset_type ---+    |      |
  |    |                                       |    | PRESSURE                 |    |      |
  |    |                                       |    | SWAT                     |    |      |
  |    |                                       |    | SGAS                     |    |      |
  |    |                                       |    +--------------------------+    |      |
  |    |    +-- local_obsset_type ---+         |                                    |      |
  |    |    | WWCT:OP_2              |         |    +--- local_dataset_type ---+    |      |
  |    |    | WGOR:OP_1              |         |    | MULTFLT1                 |    |      |
  |    |    | RFT:WELL1              |  <------|    | MULTFLT2                 |    |      |
  |    |    | RFT:WELL3              |         |    | MULTFLT3                 |    |      |
  |    |    | WWCT:WELLX             |         |    +--------------------------+    |      |
  |    |    +------------------------+         |                                    |      |
  |    |                                       |    +--- local_dataset_type ---+    |      |
  |    |                                       |    | RELPERM1                 |    |      |
  |    |                                       |    | RELPERM2                 |    |      |
  |    |                                       |    | RELPERM3                 |    |      |
  |    |                                       \    +--------------------------+    |      |
  |    |                                                                            |      |
  |    +----------------------------------------------------------------------------+      |
  |                                                                                        |
  |                                                                                        |
  |    +----------------- local_ministep_type --------------------------------------+      |
  |    |                                                                            |      |
  |    |                                       /    +--- local_dataset_type ---+    |      |
  |    |    +-- local_obsset_type ---+         |    | PERMX PORO               |    |      |
  |    |    | 4D Seismic             |         |    | PRESSURE SWAT            |    |      |
  |    |    | Gravimetri             |         |    | SGAS                     |    |      |
  |    |    |                        |  <------|    +--------------------------+    |      |
  |    |    |                        |         |                                    |      |
  |    |    |                        |         |    +--- local_dataset_type ---+    |      |
  |    |    +------------------------+         |    | MULTFLT1                 |    |      |
  |    |                                       |    | MULTFLT2                 |    |      |
  |    |                                       |    | MULTFLT3                 |    |      |
  |    |                                       \    +--------------------------+    |      |
  |    |                                                                            |      |
  |    +----------------------------------------------------------------------------+      |
  |                                                                                        |
  +----------------------------------------------------------------------------------------+

This figure illustrates the different objects when configuring local
analysis:

LocalUpdateStep: This is is the top level configuration of the
   updating at one timestep. In principle you can have different
   updatestep configurations at the different timesteps, but it will
   typically be identical for all the time steps. Observe that the
   update at one time step can typically conist of several enkf
   updates, this is handled by using several local_ministep.

local_ministep_type: The ministep defines a collection of observations
   and state/parameter variables which are mutually dependant on
   eachother and should be updated together. The local_ministep will
   consist of *ONE* local_obsset of observations, and one or more
   local_dataset of data which should be updated.

local_obsset_type: This is a collection of observation data; there is
   exactly one local_obsset for each local_ministep.

local_dataset_type: This is a collection of data/parameters which
   should be updated together in the EnKF updating.


How the local_dataset_type is configured is quite important for the
core EnKF updating:

 1. All the members in one local_dataset instance are serialized and
    packed in the A-matrix together; i.e. in the example above the
    parameters RELPERM1,RELPERM2 and RELPERM3 are updated in one go.

 2. When using the standard EnKF the X matrix is calculated using
    the actual data vectors, and the results will be identical if we
    use one large local_dataset instance or several small. However
    when using more advanced techniques where the A matrix is used
    explicitly when calculating the update this will matter.

 3. If you have not entered a local configuration explicitly the
    default ALL_ACTIVE local configuration will be used.
*/

LocalConfig::LocalConfig(const std::vector<std::string> &parameter_keys,
                         const std::vector<std::string> &obs_keys)
    : m_global_updatestep("DEFAULT"),
      m_global_ministep(
          "ALL_ACTIVE") // The strings "ALL_ACTIVE" and "ALL_OBS" are part of the API.
      ,
      m_global_obsdata("ALL_OBS") {
    for (const auto &obs_key : obs_keys)
        this->m_global_obsdata.add_node(obs_key);

    for (const auto &param_key : parameter_keys)
        this->m_global_ministep.add_active_data(param_key.c_str());

    this->m_global_ministep.add_obsdata(this->m_global_obsdata);
    this->m_global_updatestep.add_ministep(this->m_global_ministep);
}

bool LocalConfig::has_obsdata(const std::string &obs_key) const {
    return this->m_obsdata.count(obs_key) == 1;
}

LocalMinistep &LocalConfig::make_ministep(const std::string &key) {
    if (this->m_ministep.count(key) == 1)
        return this->m_ministep.at(key);

    if (!this->m_updatestep.has_value())
        this->m_updatestep = LocalUpdateStep("DEFAULT");

    this->m_ministep.emplace(key, LocalMinistep(key.c_str()));
    auto &ministep = this->m_ministep.at(key);
    this->m_updatestep->add_ministep(ministep);
    printf("-------------------------------------------------------------------------------\n");
    printf("%s:  ministep: %p  \n", __func__, &ministep);
    return ministep;
}

LocalMinistep *LocalConfig::ministep(const std::string &key) {
    if (key == this->m_global_ministep.name())
        return this->global_ministep();
    else
        return &this->m_ministep.at(key);
}

LocalMinistep *LocalConfig::global_ministep() {
    return &this->m_global_ministep;
}

LocalObsData *LocalConfig::make_obsdata(const std::string &key) {
    if (this->m_obsdata.count(key) == 1)
        throw std::invalid_argument("Tried to add existing observation key");

    this->m_obsdata.emplace(key, LocalObsData(key));
    return &this->m_obsdata.at(key);
}

LocalObsData *LocalConfig::obsdata(const std::string &key) {
    if (key == this->m_global_obsdata.name())
        return this->global_obsdata();
    else
        return &this->m_obsdata.at(key);
}

LocalObsData *LocalConfig::global_obsdata() { return &this->m_global_obsdata; }

LocalUpdateStep &LocalConfig::updatestep() {
    if (this->m_updatestep.has_value())
        return this->m_updatestep.value();
    else
        return this->m_global_updatestep;
}

void LocalConfig::clear() const {
    logger->warning(
        "The LocalConfig::clear() function is deprecated and will be removed");
}

//# semeio functions:
//# Ert.getLocalConfig()
//# LocalConfig.createObsdata()
//# LocalConfig.createMinistep()
//# LocalConfig.getUpdatestep()
//# LocalMinistep.add
//# LocalMinistep.attachObsset()
//# LocalUpdateStep.attachMinistep()

RES_LIB_SUBMODULE("local.local_config", m) {
    py::class_<LocalConfig>(m, "LocalConfig")
        .def("createMinistep", &LocalConfig::make_ministep,
             py::return_value_policy::reference_internal)
        .def("createObsdata", &LocalConfig::make_obsdata,
             py::return_value_policy::reference_internal)
        .def("getUpdatestep", &LocalConfig::updatestep,
             py::return_value_policy::reference_internal)
        .def("global_ministep", &LocalConfig::global_ministep)
        .def("global_obsdata", &LocalConfig::global_obsdata)
        .def("getMinistep", &LocalConfig::ministep)
        .def("clear", &LocalConfig::clear)
        .def("getObsdata", &LocalConfig::obsdata);
}

File: res/enkf/local_config.py
#  Copyright (C) 2012  Equinor ASA, Norway.
#
#  The file 'local_config.py' is part of ERT - Ensemble based Reservoir Tool.
#
#  ERT is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  ERT is distributed in the hope that it will be useful, but WITHOUT ANY
#  WARRANTY; without even the implied warranty of MERCHANTABILITY or
#  FITNESS FOR A PARTICULAR PURPOSE.
#
#  See the GNU General Public License at <http://www.gnu.org/licenses/gpl.html>
#  for more details.
from dataclasses import dataclass, field
from typing import List, Union, Tuple

from deprecation import deprecated
from res._lib.local.row_scaling import RowScaling

from res.enkf.local_obsdata import LocalObsdata
from res.enkf.local_ministep import LocalMinistep


@dataclass
class LocalConfig:
    ministeps: List[LocalMinistep] = field(default_factory=list)

    @deprecated(
        "Create a new LocalConfig instead",
    )
    def clear(self):
        self.ministeps = []

    def clear_active(self):
        raise NotImplementedError("This functionality has been removed, use: clear")

    @deprecated("Use add_ministep instead")
    def createMinistep(self, mini_step_key, _=None):
        """@rtype: Ministep"""
        if mini_step_key in [ministep.name() for ministep in self.ministeps]:
            raise KeyError(f"Ministep: {mini_step_key} already in local_config")
        return self.add_ministep(mini_step_key)

    @deprecated(
        "Add observations to ministep directly",
    )
    def createObsdata(self, obsdata_key):
        """@rtype: Obsdata"""
        return LocalObsdata(obsdata_key)

    @deprecated(
        "Add observations to ministep directly",
    )
    def copyObsdata(self, _, new_key):
        return LocalObsdata(new_key)

    @deprecated(
        "Loop over configuration directly",
    )
    def getUpdatestep(self):
        """@rtype: UpdateStep"""
        return self

    @deprecated(
        "Add observations to ministep directly",
    )
    def getMinistep(self, key):
        """@rtype: Ministep"""
        for ministep in self.ministeps:
            if ministep.name() == key:
                return ministep
        raise KeyError(f"No ministep named: {key}")

    @deprecated(
        "Add observations to ministep directly",
    )
    def getObsdata(self, obsdata_key):
        """@rtype: Obsdata"""
        return obsdata_key

    @deprecated("Use add_ministep")
    def attachMinistep(self, mini_step):
        if mini_step.name not in [ministep.name for ministep in self.ministeps]:
            self.ministeps.append(mini_step)

    def add_ministep(
        self,
        name,
        observations: List[Union[str, Tuple[str, List[int]]]],
        parameters: List[
            Union[str, Tuple[str, List[int]], Tuple[str, RowScaling, List[int]]]
        ],
    ):
        ministep = LocalMinistep(name)
        for observation in observations:
            if isinstance(observation, str):
                observation = [observation]
            ministep.add_observation(*observation)
        for parameter in parameters:
            if isinstance(parameter, str):
                parameter = [parameter]
            if len(parameter) > 1 and isinstance(parameter[1], RowScaling):
                ministep.add_row_scaling_parameter(*parameter)
            else:
                ministep.add_parameter(*parameter)

        self.ministeps.append(ministep)

    def __iter__(self):
        yield from self.ministeps

    def __len__(self):
        return len(self.ministeps)

    def __getitem__(self, item):
        return self.ministeps[item]

    def context_validate(
        self, valid_observations: List[str], valid_parameters: List[str]
    ) -> bool:
        errors = []
        for ministep in self.ministeps:
            for observation in ministep.observations:
                if observation not in valid_observations:
                    errors.append(
                        f"Observation: {observation} not in valid observations"
                    )
            for parameter in list(ministep.parameters.keys()) + list(
                ministep.row_scaling_parameters.keys()
            ):
                if parameter not in valid_parameters:
                    errors.append(f"Parameter: {parameter} not in valid parameters")
        if errors:
            raise ValueError(
                f"Update configuration not valid, "
                f"valid observations: {valid_observations}, "
                f"valid parameters: {valid_parameters}, errors: {errors}"
            )
